manager/controlapi/config_test.go

```diff
@@ -0,0 +1,437 @@
+
+	type listTestCase struct {
+		desc     string
+		expected []string
+		filter   *api.ListConfigsRequest_Filters
+	}
+
+	listConfigTestCases := []listTestCase{
+		{
+			desc:     "no filter: all the available configs are returned",
+			expected: allListableNames,
+			filter:   nil,
+		},
+		{
+			desc:     "searching for something that doesn't match returns an empty list",
+			expected: nil,
+			filter:   &api.ListConfigsRequest_Filters{Names: []string{"aa"}},
+		},
+		{
+			desc:     "multiple name filters are or-ed together",
+			expected: []string{"aaa", "bbb", "ccc"},
+			filter:   &api.ListConfigsRequest_Filters{Names: []string{"aaa", "bbb", "ccc"}},
+		},
+		{
+			desc:     "multiple name prefix filters are or-ed together",
+			expected: []string{"aaa", "aab", "bbb", "bbc"},
+			filter:   &api.ListConfigsRequest_Filters{NamePrefixes: []string{"aa", "bb"}},
+		},
+		{
+			desc:     "multiple ID prefix filters are or-ed together",
+			expected: []string{"aaa", "bbb"},
+			filter: &api.ListConfigsRequest_Filters{IDPrefixes: []string{
+				configNamesToID["aaa"], configNamesToID["bbb"]},
+			},
+		},
+		{
+			desc:     "name prefix, name, and ID prefix filters are or-ed together",
+			expected: []string{"aaa", "aab", "bbb", "bbc", "ccc", "ddd"},
+			filter: &api.ListConfigsRequest_Filters{
+				Names:        []string{"aaa", "ccc"},
+				NamePrefixes: []string{"aa", "bb"},
+				IDPrefixes:   []string{configNamesToID["aaa"], configNamesToID["ddd"]},
+			},
+		},
+		{
+			desc:     "all labels in the label map must be matched",
+			expected: []string{allListableNames[0], allListableNames[4], allListableNames[8]},
+			filter: &api.ListConfigsRequest_Filters{
+				Labels: map[string]string{
+					"mod2": "0",
+					"mod4": "0",
+				},
+			},
+		},
+		{
+			desc: "name prefix, name, and ID prefix filters are or-ed together, but the results must match all labels in the label map",
+			// + indicates that these would be selected with the name/id/prefix filtering, and 0/1 at the end indicate the mod2 value:
+			// +"aaa"0, +"aab"1, "abc"0, +"bbb"1, "bac"0, +"bbc"1, +"ccc"0, "cac"1, "cbc"0, +"ddd"1
+			expected: []string{"aaa", "ccc"},
+			filter: &api.ListConfigsRequest_Filters{
+				Names:        []string{"aaa", "ccc"},
+				NamePrefixes: []string{"aa", "bb"},
+				IDPrefixes:   []string{configNamesToID["aaa"], configNamesToID["ddd"]},
+				Labels: map[string]string{
+					"mod2": "0",
+				},
+			},
+		},
+	}
+
+	// ---- run the filter tests ----
+
+	for _, expectation := range listConfigTestCases {
+		result := listConfigs(&api.ListConfigsRequest{Filters: expectation.filter})
+		assert.Len(t, result, len(expectation.expected), expectation.desc)
+		for _, name := range expectation.expected {
+			assert.Contains(t, result, name, expectation.desc)
+			assert.NotNil(t, result[name], expectation.desc)
+			assert.Equal(t, configNamesToID[name], result[name].ID, expectation.desc)
```

Comment:
Either here or in the definition for `listConfigs`, do we want to assert that the data is not nil?